src/covid_model_seiir_pipeline/pipeline/forecasting/model/ode_systems/variant.py

```diff
@@ -0,0 +1,581 @@
+import numba
+import numpy as np
+
+
+N_SEIIR_COMPARTMENTS = 5
+N_WILD_TYPE_SEIIR_GROUPS = 4
+N_VARIANT_TYPE_SEIIR_GROUPS = 3
+N_IMMUNE_COMPARTMENTS = 2
+N_TRACKING_COMPARTMENTS = 13
+N_GROUPS = 2
+N_VACCINE_CATEGORIES = 5
+
+GROUP_SYSTEM_SIZE = (
+    (N_WILD_TYPE_SEIIR_GROUPS + N_VARIANT_TYPE_SEIIR_GROUPS) * N_SEIIR_COMPARTMENTS
+    + N_IMMUNE_COMPARTMENTS
+    + N_TRACKING_COMPARTMENTS
+)
+N_VACCINE_PARAMETERS = N_VACCINE_CATEGORIES * N_GROUPS
+
+# 3rd and 4th compartment of each seiir group are infectious.
+LOCAL_I1 = 2
+LOCAL_I2 = 3
+
+COMPARTMENTS = [
+    'S',            'E',            'I1',            'I2',            'R',             # Unvaccinated
+    'S_u',          'E_u',          'I1_u',          'I2_u',          'R_u',           # Vaccinated and unprotected
+    'S_p',          'E_p',          'I1_p',          'I2_p',          'R_p',           # Vaccinated and protected wild type only
+    'S_pa',         'E_pa',         'I1_pa',         'I2_pa',         'R_pa',          # Vaccinated and protected all
+    'S_variant',    'E_variant',    'I1_variant',    'I2_variant',    'R_variant',     # Unvaccinated
+    'S_variant_u',  'E_variant_u',  'I1_variant_u',  'I2_variant_u',  'R_variant_u',   # Vaccinated and unprotected
+    'S_variant_pa', 'E_variant_pa', 'I1_variant_pa', 'I2_variant_pa', 'R_variant_pa',  # Vaccinated and protected all
+    'S_m',                                                                             # Immunized wild type only
+    'R_m',                                                                             # Immunized all
+    'NewE_wild', 'NewE_variant', 'NewE_p_wild', 'NewE_p_variant',                      # Tracking compartments.
+    'NewE_nbt', 'NewE_vbt', 'NewS_v', 'NewR_w',
+    'V_u', 'V_p', 'V_pa', 'V_m', 'V_ma',
+]
+
+##############################################
+# Give indices semantically meaningful names #
+##############################################
+# Parameters
+alpha, beta, beta_b117, beta_b1351, sigma, gamma1, gamma2 = 0, 1, 2, 3, 4, 5, 6
+theta_plus, theta_minus = 7, 8
+b117_prevalence, b1351_prevalence = 9, 10
+p_cross_immune = 11
+
+# Compartments
+# See ordering in /lib/static_vars.py
+s, e, i1, i2, r = 0, 1, 2, 3, 4
+s_u, e_u, i1_u, i2_u, r_u = 5, 6, 7, 8, 9
+s_p, e_p, i1_p, i2_p, r_p = 10, 11, 12, 13, 14
+s_pa, e_pa, i1_pa, i2_pa, r_pa = 15, 16, 17, 18, 19
+
+s_variant, e_variant, i1_variant, i2_variant, r_variant = 20, 21, 22, 23, 24
+s_u_variant, e_u_variant, i1_u_variant, i2_u_variant, r_u_variant = 25, 26, 27, 28, 29
+s_pa_variant, e_pa_variant, i1_pa_variant, i2_pa_variant, r_pa_variant = 30, 31, 32, 33, 34
+
+s_m, r_ma = 35, 36
+
+# Composite tracking compartments.
+new_e_wild, new_e_variant, new_e_p_wild, new_e_p_variant = 37, 38, 39, 40
+new_e_nbt, new_e_vbt, new_s_v, new_r_w = 41, 42, 43, 44
+v_u, v_p, v_pa, v_m, v_ma = 45, 46, 47, 48, 49
+
+# Vaccine categories.
+u, p, pa, m, ma = 0, 1, 2, 3, 4
```

Comment:
Suggestion: destructure COMPARTMENTS

```python
# massive destructuring - keeps us honest
(
    s, e, i1, i2, r ,
    s_u, e_u, i1_u, i2_u, r_u ,
    s_p, e_p, i1_p, i2_p, r_p ,
    s_pa, e_pa, i1_pa, i2_pa, r_pa ,

    s_variant, e_variant, i1_variant, i2_variant, r_variant ,
    s_u_variant, e_u_variant, i1_u_variant, i2_u_variant, r_u_variant ,
    s_pa_variant, e_pa_variant, i1_pa_variant, i2_pa_variant, r_pa_variant ,

    s_m, r_ma ,

    # Composite tracking compartments.
    new_e_wild, new_e_variant, new_e_p_wild, new_e_p_variant ,
    new_e_nbt, new_e_vbt, new_s_v, new_r_w ,
    v_u, v_p, v_pa, v_m, v_ma ,
) = range(len(COMPARTMENTS))

# proof I'm not betraying you
print(s)
print(s_m)
print(new_e_nbt)
print(v_ma)
```

```shell
$ python big_destructure.py 
0
35
41
49
```

Reply:
This is great.  I tried a few abstractions but they broke numba. This seems perfect.